File: javascript/ledger/ledger.js
// Change Log:
// * Modified Ledger Entry constructor to take date, descritpion, and change that way it is a one-liner
//   and not a four liner. Later date format fixing added some lines back in.
// * Wanted to remove function createEntry as it is the same amount of code to just make a new LedgerEntry.
//   Unfortunately, the test cases need it.
// * Added T00:00 to date values when needed so they would parse correctly when parsed during new Date(date)
//   how did this work without that?
// * formatEntries move header to top pulls locale specific info from a locale dictionary
// * looks like formatEntries sort function is the same either way, move that out. 
//   I put it in ledgers dictionary so that other locales may sort differently if added.
// * Change date formatting custom code to use .toLocaleDateString() with an options object to get two digt month/year
// * Change currency format options to also pull from the locales dictionary. Didn't work, moved to a currency dictionary
//   that still didn't work either so, both locale and currency for a key. Went for both locale and currency as
//   new keys, and finally got that to work. Not sure if NL-nl EUR should have parenthesized negative number or not (No test for that).
// * With all that out of the way we don't need to have code specific to locale in the write records loop so it was
//   simplified to one loop for everything. Trimming for negative numbers did get strange.


class LedgerEntry {
  constructor(date, description, change) {
    this.date = date;
    this.description = description;
    this.change = change;
  }
}

export const createEntry = (date, description, change) => {  
  let dateStr = date
  if (date.length <= 10 && date.indexOf('T') < 0) 
    dateStr += 'T00:00'
  return new LedgerEntry(new Date(dateStr), description, change)
}

let ledgerEntrySort = (a, b) => {
  return a.date - b.date ||
    a.change - b.change ||
    a.description.localeCompare(b.description)
}

const currencies = {
  'en-US_USD': {
    style: 'currency',
    currency: 'USD',
    currencyDisplay: 'narrowSymbol',
    currencySign: 'accounting',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  },
  'en-US_EUR': {
    style: 'currency',
    currency: 'EUR',
    currencyDisplay: 'narrowSymbol',
    currencySign: 'accounting',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  },
  'nl-NL_USD': {
    style: 'currency',
    currency: 'USD',
    currencyDisplay: 'narrowSymbol',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  },
  'nl-NL_EUR': {
    style: 'currency',
    currency: 'EUR',
    currencyDisplay: 'narrowSymbol',
    currencySign: 'accounting',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }
}

const locales= {
  'en-US': {
    locale: 'en-US', 
    date: 'Date', 
    description: 'Description', 
    change: 'Change', 
    sort: ledgerEntrySort, 
  },
  'nl-NL': {
    locale: 'nl-NL', 
    date: 'Datum', 
    description: 'Omschrijving', 
    change: 'Verandering', 
    sort: ledgerEntrySort,
  }
}

export function formatEntries(currency, locale, entries) {
  if (!(locale in locales))
  {
    throw new Error(`Locale: ${locale} is not supported.`)
  }

  let currencyKey = `${locale}_${currency}` 
  if (!(currencyKey in currencies))
  {
    throw new Error(`Locale: ${locale}, and Currency: ${currency} combination is not supported.`)
  }
      
  let table = '';
  // Generate Header Row
  table += `${locales[locale].date.padEnd(10, ' ')} | ${locales[locale].description.padEnd(25, ' ')} | ${locales[locale].change.padEnd(13, ' ')}\n`  

  // Sort entries
  entries.sort(locales[locale].sort)
  
  let dateFormatOptions = { year: 'numeric', day: '2-digit', month: '2-digit'}
  entries.forEach((entry) => {
    // Write entry date to table
    const dateStr = entry.date.toLocaleDateString(locale, dateFormatOptions)
    table += `${dateStr} | `;

    // Write entry description to table
    const truncatedDescription =
      entry.description.length > 25
        ? `${entry.description.substring(0, 22)}...`
        : entry.description.padEnd(25, ' ');
    table += `${truncatedDescription} | `;

    // Write entry change to table
    
    let shouldTrim = entry.change < 0 && 
          currencies[currencyKey].currencySign === 'accounting'
    let changeStr = `${(entry.change / 100).toLocaleString(
          locale,
          currencies[currencyKey],
        )} `;
    if (shouldTrim)
        changeStr = changeStr.trim()
    table += changeStr.padStart(13, ' ');
    table += '\n';
  });
  
  return table.replace(/\n$/, '');
}

File: python/ledger/ledger.py
"""
Ledger Printer

Change Log:
* Add comments that the test suite will complain about if missing.
* Want to merge LedgerEntry.__init__, and create_entry. Can't do that, the test suite needs it. Still moved code into the constructor.
* Reduced header line in format_entries to a single line of code.
* Write date is now a one liner
* Write description is now a one liner
* Currency required the locale library and a special function but it is shorter now. The formatting doesn't match what the locale expects, it is off and that is what required a custom function.
* Both locales handled in one branch removing duplicated code.
* Sort the entries and iterate over them instead of finding the minimum value and popping each time.
* Exercism is broken and doesn't behave the same locally and on their server. Had to remove the locale library and roll my own thing even more.
"""
from datetime import datetime

# Currency code to symbol
# Could add more currencies this way
currency_lookup = {
    'USD': '$',
    'EUR': '€',
}
language_lookup = {'en_US':  {
        'Date': 'Date',
        'Description': 'Description',
        'Change': 'Change',
        'DateFormat': '%m/%d/%Y',
        'Decimal': '.',
        'Thousands': ','
    },
    'nl_NL': {
        'Date': 'Datum',
        'Description': 'Omschrijving',
        'Change': 'Verandering',
        'DateFormat': '%d-%m-%Y',
        'Decimal': ',', 
        'Thousands': '.'
    }
}


class LedgerEntry:
    """
    Ledger Entry class (describes on ledger entry)
    This one is going to throw a too few methods error.
    """
    def __init__(self, date, description, change):
        """
        Initialze Ledger Entry to default values
        :param date: Date of the entry
        :param description: Description of the entry
        :param change: monetary amount
        """
        self.date = datetime.strptime(date, '%Y-%m-%d')
        self.description = description
        self.change = change


def create_entry(date, description, change):
    """Create a LedgerEntry object with the given properties
    :param date: Date of the transaction
    :param description: Description of the transaction
    :change: Amount of the transaction (implied 2 digit decimal)
    """
    return LedgerEntry(date, description,change)

def format_currency(currency, currency_locale, amount):
    """Format Currency for the desired locale and currency
    :param currency: Currency of the transaction USD or EUR
    :param currency_locale: Locale of the transaction en_US or nl_NL
    :param amount: How much the transaction was for
    :returns: The currency amount formatted properly.
    """
    # write the number as a string for the desired locale 
    # will add () for negative numbers if en_US so give it ABS() if en_US locale. Don't want two negative indicators after all.
    symbol = currency_lookup[currency]
    symbol = (symbol if currency_locale == 'en_US' else symbol + ' ')
    value = f"{symbol}{abs(amount) if currency_locale == 'en_US' else amount:_.2f}"
    value = value.replace('.', language_lookup[currency_locale]['Decimal']).replace('_', language_lookup[currency_locale]['Thousands'])
    # Negative numbers get parenthesis in the US. If no parenthesis add an extra space so everything lines up
    if amount < 0 and currency_locale == 'en_US':
        value = '(' + value + ')'
    else:
        value += ' '
    return value


def format_entries(currency, current_locale, entries):
    """
    Write out the entries in the list as a formatted table.
    :param currency: Monetary currency 'USD' or 'EUR'
    :param current_locale: Locale of the transaction 'en_US' or 'nl_NL'
    :entries: List of LedgerEntry objects
    :returns: string with the entries nicely formatted in a table.
    """

    # Generate Header Row
    table = language_lookup[current_locale]['Date'].ljust(11) +  '| ' + \
    language_lookup[current_locale]['Description'].ljust(26) +  '| ' + \
    language_lookup[current_locale]['Change'].ljust(13)

    sorted_entries = sorted(entries, key=lambda x: (x.date, x.change, x.description))

    for entry in sorted_entries:
        table += '\n'
        # Write entry date to table
        table += entry.date.strftime(language_lookup[current_locale]['DateFormat']) 
        table += ' | '

        # Write entry description to table
        # Truncate if necessary
        table += entry.description[:22] + '...' if len(entry.description) > 25 else entry.description.ljust(25)
        table += ' | '

        # Write entry change to table
        table += format_currency(currency, current_locale, entry.change / 100).rjust(13)
    return table
